### PR TITLE
Add async iterator blog

### PR DESCRIPTION
**Pull request by bterlson**
I'll continue to noodle on this (generally cleaning things up, adding more realistic samples, and trimming content where I can), but getting this out there for feedback!

### CODE REVIEW

File: _posts/2020-05-26-async-iterators.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+## Background
+
+At a high level, a function might produce one value or a (possibly unending) sequence of values. It also might do its work synchronously or asynchronously. Considered together, these modes create four possible ways functions might be declared and consumed. Put in a table, it looks like the following:
+
+|                  | One Value                             | Many Values                                   |
+| ---------------- | ------------------------------------- | --------------------------------------------- |
+| **Synchronous**  | function<br>Returns T                 | generator function<br>Returns Iterator\<T>    |
+| **Asynchronous** | async function<br>Returns Promise\<T> | async generator <br>Returns AsyncIterator\<T> |
+
+When JavaScript debuted in 1995, it supported only synchronous functions that produce only a single return value. Developers who wanted to return multiple values, or wanted to produce values asynchronously, had to rely on libraries (for example, node's Stream or promises).
```
Comment (xirzec) [inline _posts/2020-05-26-async-iterators.md:23]:
```suggestion
When JavaScript debuted in 1995, it supported only synchronous functions that produce only a single return value. Developers who wanted to return multiple values, or wanted to produce values asynchronously, had to rely on libraries (for example, node's Stream or promises.)
```

Comment (xirzec) [reply _posts/2020-05-26-async-iterators.md:23]: also is 'node' always supposed to be lowercase like npm? Feels weird that Stream is capitalized but not Promise? (Since Promise is still a constructor, yeah?)

File: _posts/2020-05-26-async-iterators.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  };
+}
+
+const iterator1 = counter(3);
+iterator1.next(); // { value: 0, done: false }
+
+const iterator2 = counter(3);
+iterator2.next(); // { value: 0, done: false }
+```
+
```
Comment (xirzec) [inline _posts/2020-05-26-async-iterators.md:75]: I feel like there needs to be some text between these two samples? I get a bit lost as to why we go back to the fixed sized iterator

File: _posts/2020-05-26-async-iterators.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      return { value: this.count, done: true };
+    },
+
+    count: 0,
+  };
+}
+```
+
+Usually, however, we have an object that isn't itself an iterator with all its state, but instead iterable in the sense that it can vend fresh iterators over some sequence. So we could wrap the iterator above in a function that returns a new iterator each time its called:
+
```
Comment (xirzec) [inline _posts/2020-05-26-async-iterators.md:142]: missing example?

File: _posts/2020-05-26-async-iterators.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+|                  | One Value                             | Many Values                                   |
+| ---------------- | ------------------------------------- | --------------------------------------------- |
+| **Synchronous**  | function<br>Returns T                 | generator function<br>Returns Iterator\<T>    |
+| **Asynchronous** | async function<br>Returns Promise\<T> | async generator <br>Returns AsyncIterator\<T> |
+
+When JavaScript debuted in 1995, it supported only synchronous functions that produce only a single return value. Developers who wanted to return multiple values, or wanted to produce values asynchronously, had to rely on libraries (for example, node's Stream or promises).
+
+### Iterators & Generators
+
+The 6th edition of the JavaScript language, ES2015, added iterators and iterables. Iterables are any object with a `Symbol.iterator` method that returns an iterator. Iterators are objects with a `next()` method. Calling `next()` returns an object with `value` and `done` properties which tell you the current iteration value and whether more values are available.
```
Comment (xirzec) [inline _posts/2020-05-26-async-iterators.md:27]: I think it would be good to add some motivation here by showing the nice for/of syntax that works against any iterable (providing a standard way to iterate many types!)

File: _posts/2020-05-26-async-iterators.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    },
+
+    count: 0,
+  };
+}
+```
+
+Usually, however, we have an object that isn't itself an iterator with all its state, but instead iterable in the sense that it can vend fresh iterators over some sequence. So we could wrap the iterator above in a function that returns a new iterator each time its called:
+
+ES2015 also added Promises to the language, although without any syntactic support. The 8th edition of the JavaScript language, ES2017, added `async function`s which offered developers an easy way to define functions that returned promises. Additionally, developers could `await` any promise value inside the async function effectively deferring further execution of the function until the awaited promise resolves successfully. This made consuming asynchronous APIs substantially easier than dealing with promises directly - no more `then()`/`catch()` calls!
```
Comment (xirzec) [inline _posts/2020-05-26-async-iterators.md:143]: should we show an async iterator and for await of?

File: _posts/2020-05-26-async-iterators.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
+---
+title: AsyncIterators in the Azure SDK
+layout: post
+date: 26 may 2020
```
Comment (adrianhall) [inline _posts/2020-05-26-async-iterators.md:4]: Date is wrong format - needs to be YYYY-MM-DD (same as filename).  Also, please update to new date.

File: _posts/2020-05-26-async-iterators.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+sidebar: releases_sidebar
+author_github: bterlson
+repository: azure/azure-sdk
+---
+
+Async Iterators represent a stream of data that is loaded asynchronously. They are used across the Azure SDK to represent asynchronous streams of data such as paginated APIs where data is fetched a page at a time.
```
Comment (adrianhall) [inline _posts/2020-05-26-async-iterators.md:10]: I think we probably need to note it's for JavaScript early on and in the title.

Comment (adrianhall) [inline _posts/2020-05-26-async-iterators.md:10]: potentially link to the MDN spec?

File: _posts/2020-05-26-async-iterators.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+Async Iterators were added fairly recently to JavaScript - the ES2018 edition of the standard. They are supported natively in recent version of modern browsers, but TypeScript can compile them down to generators or even vanilla functions.
```
Comment (adrianhall) [inline _posts/2020-05-26-async-iterators.md:12]: Link to the spec.

Comment (adrianhall) [inline _posts/2020-05-26-async-iterators.md:12]: Also, link TypeScript.... to the typescriptlang website entry on them

Comment (adrianhall) [reply _posts/2020-05-26-async-iterators.md:12]: Also, most devs don't get generators, so I'd reword this to say that TypeScript can compile them into code that can run on prior versions of JavaScript.

File: _posts/2020-05-26-async-iterators.md (hunk 5 above)
Comment (adrianhall) [inline _posts/2020-05-26-async-iterators.md:143]: Needs the closing content - For more information / follow us on twitter content. (see a previous blog post)

File: _posts/2020-05-26-async-iterators.md (hunk 1 above)
Comment (bterlson) [reply _posts/2020-05-26-async-iterators.md:23]: I don't agree with many style guide's placement of periods inside parens and will not take that suggestion 😁

Comment (bterlson) [reply _posts/2020-05-26-async-iterators.md:23]: `node` or `Node.js` are both acceptable according to node.

node's Stream is a specific implementation/type. Promises refer to any number of libraries and even encapsulates e.g. jquery deferred's, and so I went with lowercase. But I see your point.

File: _posts/2020-05-26-async-iterators.md (hunk 2 above)
Comment (bterlson) [reply _posts/2020-05-26-async-iterators.md:75]: The counter example was detritus and is now removed.

File: _posts/2020-05-26-async-iterators.md (hunk 5 above)
Comment (bterlson) [reply _posts/2020-05-26-async-iterators.md:143]: This is showed in detail later.

File: _posts/2020-05-26-async-iterators.md (hunk 6 above)
Comment (bterlson) [reply _posts/2020-05-26-async-iterators.md:4]: hmm I used this format in the last version (which I copied from I think a Java blog) but I'm always happy to use ISO-8601 formatted dates.

File: _posts/2020-05-26-async-iterators.md (hunk 7 above)
Comment (bterlson) [reply _posts/2020-05-26-async-iterators.md]: I wish. There is no good documentation page on MDN for this believe it or not.

Comment (bterlson) [reply _posts/2020-05-26-async-iterators.md]: Clarified this a bit.

**Comment by bterlson** [comment]
Ok I think I addressed all your feedback @adrianhall. If you can, another look would be great, and I can make any minor tweaks on monday!
